docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,211 @@
+
+Whales can't easily redelegate their positions, since they do not want to lose
+out on an epoch of rewards.
+
+2. Rewards calculations at epoch boundaries take too long
+
+Since all of the stake accounts simultaneously receive rewards at the start of an epoch,
+the network experiences a big load to process these accounts. This feature
+is extremely useful for users, since they don't need to do anything to claim
+rewards. On the flip side, it introduces too much load on validators.
+
+With a new stake account, we can address both of these issues.
+
+## New Stake Account
+
+The current `StakeState` is 200 bytes, and can be in 4 states:
+
+* `Uninitialized`
+* `Initialized(Meta)`: the rent-exempt reserve, authorized keys, and lockup
+* `Stake(Meta, Stake)`: `Delegation` (validator, stake amount, activation / deactivation epochs, warmup rate) + `credits_observed` (used for rewards calc)
+* `RewardsPool`: deprecated (?)
+
+To properly perform a redelegation, we need to store at least the
+next validator's vote pubkey and a redelegation epoch, which means that
+redelegation absolutely requires a larger struct.
+
+### Structure
+
+We need to introduce a new type and an upgrade path from legacy / V1 stakes:
+
+```
+struct StakeV2 {
+    meta: Meta,
+    stake: Stake,
+    last_claim_epoch: Epoch, // needed to know if it's up-to-date, see `ClaimRewards`
+    redelegation: Redelegation,
+}
+struct Redelegation {
+    voter_pubkey: Pubkey,
+    redelegation_epoch: Epoch,
+    credits_observed: u64,
+}
+```
+
+With this information, we can calculate multi-epoch redelegations when the cluster
+sees more than 25% stake movement.
+
+### Interoperability with V1
+
+Since loads of places use `mem::size_of::<StakeState>()`, and we still need
+to support that, that struct won't get touched. On the flip side, programs
+need to process both. Since there's already a neat interface on `StakeState`,
+we can do something similar with a new type:
+
+```
+enum StakeStateV2 {
+    Uninitialized,
+    Initialized(Meta),
+    StakeV1(Meta, Stake),
+    RewardsPool,
+    StakeV2(StakeV2),
+}
+impl StakeStateV2 {
+    /// ... all the same as StakeState ...
+    pub fn redelegation(&self) -> Option<Redelegation> {
+      match self {
+          Self::StakeV2(stake) => Some(stake.redelegation),
+          _ => None,
+      }
+    }
+}
+```
+
+Programs update to use `StakeStateV2` everywhere, and existing code still works.
+When an `Initialized` stake is delegated, the stake program makes it a `StakeV1`
+if the size is `200`, and a `StakeV2` if the size is `256`.
+
+### Epoch Boundary Calculations
+
+While we're at it, new stake accounts can also solve point number 2, by having
+their rewards paid through an update instruction, and not automatically at every
+epoch boundary.
+
+During epoch boundary, the validator vote accounts receive all stake V2
+account rewards as withheld lamports in their account. They cannot
+touch these, and stake accounts may claim them.
```

Comment:
how would this scheme handle compounding stake if the delegator leaves rewards unclaimed for multiple epochs?  seems like we lose the proportionality, so even if we count these lamports as global stake (reasonable), we won't know how to distribute it.